Initial commit for multiprocessing firewall engine

VTabolin opened the pull request.

VTabolin committed:
Initial commit for multiprocessing firewall engine

ekosareva commented on vmware_dvs/agent/dvs_neutron_agent.py:
```diff
@@ -65,7 +65,6 @@ def __init__(self, vsphere_hostname, vsphere_login, vsphere_password,
             'start_flag': True}
 
-        self.setup_rpc()
         report_interval = cfg.CONF.AGENT.report_interval
```
_report_state should be started after all rpc agents were initialized. So, please move L68-L72 under setup_rpc() function

VTabolin replied:
Done

VTabolin committed:
Port bounding lookup works faster, update for firewall engine

New version of lookup_unbound_port function implemented. Improved
stability for multiprocessing firewall. Queue balancer added

ekosareva commented on vmware_dvs/agent/firewalls/vcenter_firewall.py:
```diff
@@ -26,6 +31,56 @@
+            time.sleep(1)
+
+
+class port_queue(object):
```
Regarding pep8 class names should normally use the CapWords convention. Please, change port_queue -> PortQueue

ekosareva commented on vmware_dvs/agent/firewalls/vcenter_firewall.py:
```diff
@@ -26,6 +31,58 @@
+    def get_tasks(self, number=0):
+        for dvs in self.taskstore:
+            tasks = self.taskstore.get(dvs, [])
+            if len(tasks) > 0:
```
lines 54-56 is equal with:
for dvs, tasks in self.taskstore.iteritems():
    if tasks:

ekosareva commented on vmware_dvs/agent/firewalls/vcenter_firewall.py:
```diff
@@ -75,7 +138,29 @@ def _apply_sg_rules_for_port(self, ports):
+                    for port in ports:
+                        queue = self._getfreequeue()
+                        print queue
+                        queue.put({physnet: [port]})
```
please, rewrite lines 143-150:

```
for port in port_list:
    queue = self._getfreequeue()
    queue.put({
        physnet: [{'id': port['id'],
                   'security_group_rules': port['security_group_rules'],
                   'binding:vif_details': port['binding:vif_details']}]
    })
```

ekosareva commented on vmware_dvs/agent/firewalls/vcenter_firewall.py:
```diff
@@ -75,7 +138,29 @@ def _apply_sg_rules_for_port(self, ports):
+                if minimum == 0:
+                    return queue
+                ret = queue
+        return ret
```
```
shortest_queue = self.list_queues[0]
for queue in self.list_queues:
    queue_size = queue.qsize()
    if queue_size == 0:
        return queue
    if queue_size < shortest_queue.qsize():
        shortest_queue = queue
return shortest_queue         
```

ekosareva commented on vmware_dvs/utils/dvs_util.py:
```diff
@@ -325,6 +325,27 @@ def _get_ports_for_pg(self, pg_name):
     def _lookup_unbound_port(self, port_group):
         builder = SpecBuilder(self.connection.vim.client.factory)
+        criteria = builder.port_criteria(port_group_key=port_group.value,
+                                         connected=None)
```
It's not obligatory pass `connected` param, by default connected=None

ekosareva commented on vmware_dvs/utils/dvs_util.py:
```diff
@@ -325,6 +325,27 @@ def _get_ports_for_pg(self, pg_name):
+        port_keys = [x for x in free_keys if x not in self._blocked_ports]
+        if len(port_keys) > 0:
+            self._blocked_ports.add(port_keys[0])
+            return self._get_port_info_by_portkey(port_keys[0])
```
may be rewrite lines 340-345:

```
        for port_key in all_port_keys:
            if (port_key not in connected_port_keys and 
                    port_key not in self._blocked_ports):
                self._blocked_ports.add(port_key)
                return self._get_port_info_by_portkey(port_key)  
        raise exceptions.UnboundPortNotFound()
```

ekosareva commented on vmware_dvs/agent/firewalls/vcenter_firewall.py:
```diff
@@ -13,6 +13,11 @@
 #    under the License.
 
+from multiprocessing import Process, Queue
+
```
this empty line is not required

VTabolin committed:
Remove ports queue added for async firewall engine
